substrate-node/pallets/pallet-smart-contract/src/migrations/v7.rs

```diff
@@ -0,0 +1,135 @@
+use codec::{Decode, Encode};
+#[cfg(feature = "try-runtime")]
+use sp_std::vec::Vec;
+
+pub struct FixTwinLockedBalances<T: Config>(PhantomData<T>);
+
+impl<T: Config> OnRuntimeUpgrade for FixTwinLockedBalances<T> {
+    #[cfg(feature = "try-runtime")]
+    fn pre_upgrade() -> Result<Vec<u8>, &'static str> {
+        info!("current pallet version: {:?}", PalletVersion::<T>::get());
+        assert!(PalletVersion::<T>::get() >= types::StorageVersion::V6);
+
+        info!("👥  Smart Contract pallet to V7 passes PRE migrate checks ✅",);
+        Ok(vec![])
+    }
+
+    fn on_runtime_upgrade() -> Weight {
+        if PalletVersion::<T>::get() == types::StorageVersion::V6 {
+            migrate_to_version_7::<T>()
+        } else {
+            info!(" >>> Unused Smart Contract pallet V7 migration");
+            Weight::zero()
+        }
+    }
+
+    #[cfg(feature = "try-runtime")]
+    fn post_upgrade(pre_contracts_count: Vec<u8>) -> Result<(), &'static str> {
+        info!("current pallet version: {:?}", PalletVersion::<T>::get());
+        assert!(PalletVersion::<T>::get() >= types::StorageVersion::V7);
+
+        info!(
+            "👥  Smart Contract pallet to {:?} passes POST migrate checks ✅",
+            PalletVersion::<T>::get()
+        );
+
+        Ok(())
+    }
+}
+
+pub fn migrate_to_version_7<T: Config>() -> frame_support::weights::Weight {
+    info!(
+        " >>> Starting contract pallet migration, pallet version: {:?}",
+        PalletVersion::<T>::get()
+    );
+
+    let mut read_writes = 0;
+
+    let mut twin_contract_locked_balances: BTreeMap<u32, BalanceOf<T>> = BTreeMap::new();
+
+    for (ctr_id, l) in ContractLock::<T>::iter() {
+        let ctr = Contracts::<T>::get(ctr_id);
+        read_writes += 1;
+        match ctr {
+            Some(contract) => {
+                // l.amount_locked
+                twin_contract_locked_balances
+                    .entry(contract.twin_id)
+                    .and_modify(|v| *v += l.amount_locked)
+                    .or_insert(BalanceOf::<T>::saturated_from(0 as u128));
+            }
+            None => (),
+        }
+    }
+
+    for (t, total_contract_locked_balance) in twin_contract_locked_balances {
+        let twin = pallet_tfgrid::Twins::<T>::get(t);
+        read_writes += 1;
+        match twin {
+            Some(twin) => {
+                let total_lock_balances = get_locked_balance::<T>(&twin.account_id);
+
+                if total_lock_balances != total_contract_locked_balance {
+                    log::info!(
+                        "total locked balance on twin {} account: {:?}",
+                        t,
+                        total_lock_balances
+                    );
+                    log::info!(
+                        "should have locked only: {:?}",
+                        total_contract_locked_balance
+                    );
+
+                    let total_balance = <T as Config>::Currency::total_balance(&twin.account_id)
```

Comment:
Just to clarify a comment here would be good

I am confused  between `Currency::total_balance()` and `Currency::free_balance()` ?

and what is exactly `Currency::minimum_balance()`?

Reply:
- `total_balance` is the total amount of balance on the account (excluding locks)
- `free_balance` is the total free balance
- `minimum_balance` is the existince requirement, an account cannot fall below this amount of balance or it's "destroyed".